Fix DAG bundle imports in subprocess operators

gfreitash opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

**closes: #56783**

Resolve `ModuleNotFoundError` for DAG bundle imports within standard operator subprocesses in Airflow 3.x.

## Problem
Subprocesses spawned by these operators don't have the DAG bundle path in `PYTHONPATH` (see #53617), causing imports to fail when modules from the bundle are imported inside the callable function.

## Solution
Extracts bundle path from task instance context during execution call
Add extracted bundle path to subprocess `PYTHONPATH` environment variable
Guard with `AIRFLOW_V_3_0_PLUS` check for backward compatibility

gfreitash committed:
Fix DAG bundle imports in subprocess operators (#56783)

Add the bundle path to PYTHONPATH in operators on the standard provider that runs in a subprocess to fix import errors in Airflow 3.x.

gfreitash committed:
Apply precommit formatting changes

potiuk reviewed (approved):
It looks great. I would love however that someone else from the bundle-work-related commiters take a look.

@ephraimbuddy @jedcunningham ? Maybe one of you :) ?

ephraimbuddy commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -1189,7 +1258,7 @@ def f():
 @pytest.mark.execution_timeout(120)
 @pytest.mark.virtualenv_operator
-class TestPythonVirtualenvOperator(BaseTestPythonVirtualenvOperator):
+class TestPythonVirtualenvOperator(BaseTestPythonVirtualenvOperator, _TestDagBundleImportMixin):
```
I will prefer a separate test with the Mixin instead of adding it to the existing tests

gfreitash replied:
Changed it to a dedicated class instead of mixin + inheritance

ephraimbuddy commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -933,6 +933,23 @@ def poke(self, context):
+        from bug_test_dag_repro.lib.helper import get_message
+
+        return get_message()
+    except Exception as e:
```
```suggestion
    except ImportError as e:
```

gfreitash committed:
Change broad exception to specific ImportError in DAG bundle test helper

The _callable_that_imports_from_bundle function was catching broad
Exception which could mask unexpected errors. Changed to catch
ImportError specifically since that's what we're testing for.

gfreitash committed:
Refactor DAG bundle import tests into separate parameterized test class

Move DAG bundle import tests from mixin pattern to a dedicated parameterized test class
to improve test isolation. The new TestDagBundleImportInSubprocess
class tests subprocess-based Python operators using pytest parametrization.

gfreitash committed:
Fix CI ruff static check, parametrize attribute expects tuple instead of plain string on the first argument

jason810496 reviewed (approved):
Thanks for the PR! LGTM overall.

Lee-W commented on providers/standard/src/airflow/providers/standard/operators/python.py:
```diff
@@ -488,8 +488,30 @@ def execute(self, context: Context) -> Any:
+            if bundle_instance and hasattr(bundle_instance, "path"):
+                return Path(bundle_instance.path)
+
+        return None
```
```suggestion
        if (
            ti := context.get("ti")
            and bundle_instance := getatter(ti, "bundle_instance", None)
            and path := getattr(bundle_instance, "path", None)
        ):
            return Path(path)
            
        return None    
```

ashb replied:
ti can never be null - this code runs in a task only

gfreitash replied:
I simplified it to just `context["ti"].bundle_instance.path`, as right before it I check for Airflow 3 and as you said, `ti` can't be null

Lee-W commented on providers/standard/src/airflow/providers/standard/operators/python.py:
```diff
@@ -565,6 +587,18 @@ def _execute_python_callable_in_subprocess(self, python_path: Path):
             if self.env_vars:
                 env_vars.update(self.env_vars)
 
+            # Add bundle_path to PYTHONPATH for subprocess to import DAG bundle modules
```
```suggestion
            # Add bundle_path to PYTHONPATH for subprocess to import Dag bundle modules
```

Lee-W commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -937,6 +937,88 @@ def poke(self, context):
 
+def _callable_that_imports_from_bundle():
+    """
+    A callable function for testing DAG bundle imports.
```
```suggestion
    A callable function for testing Dag bundle imports.
```

Lee-W commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -937,6 +937,88 @@ def poke(self, context):
+)
+class TestDagBundleImportInSubprocess(BasePythonTest):
+    """
+    Test DAG bundle imports for subprocess-based Python operators.
```
```suggestion
    Test Dag bundle imports for subprocess-based Python operators.
```

Lee-W commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -937,6 +937,88 @@ def poke(self, context):
+    Test DAG bundle imports for subprocess-based Python operators.
+
+    This test ensures that callables running in subprocesses can import modules
+    from their DAG bundle by verifying PYTHONPATH is correctly set (Airflow 3.x+).
```
```suggestion
    from their Dag bundle by verifying PYTHONPATH is correctly set (Airflow 3.x+).
```

Lee-W commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -937,6 +937,88 @@ def poke(self, context):
+    from their DAG bundle by verifying PYTHONPATH is correctly set (Airflow 3.x+).
+    """
+
+    @pytest.mark.skipif(not AIRFLOW_V_3_0_PLUS, reason="DAG Bundle import fix is for Airflow 3.x+")
```
```suggestion
    @pytest.mark.skipif(not AIRFLOW_V_3_0_PLUS, reason="Dag Bundle import fix is for Airflow 3.x+")
```

Lee-W commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -937,6 +937,88 @@ def poke(self, context):
+    def test_dag_bundle_import_in_subprocess(self, dag_maker, opcls, pytest_marks, test_class_ref):
+        """
+        Tests that a callable in a subprocess can import modules from its
+        own DAG bundle (fix for Airflow 3.x).
```
```suggestion
        own Dag bundle (fix for Airflow 3.x).
```

Lee-W commented on providers/standard/tests/unit/standard/operators/test_python.py:
```diff
@@ -937,6 +937,88 @@ def poke(self, context):
+        Tests that a callable in a subprocess can import modules from its
+        own DAG bundle (fix for Airflow 3.x).
+        """
+        with TemporaryDirectory() as tmp_dir:
```
Should we just use https://docs.pytest.org/en/stable/how-to/tmp_path.html#the-tmp-path-fixture

ashb replied:
Yes

ashb commented on providers/standard/src/airflow/providers/standard/operators/python.py:
```diff
@@ -565,6 +587,18 @@ def _execute_python_callable_in_subprocess(self, python_path: Path):
+
+                if existing_pythonpath:
+                    # Prepend bundle_path to existing PYTHONPATH
+                    env_vars["PYTHONPATH"] = f"{bundle_path_str}{path_separator}{existing_pythonpath}"
```
```suggestion

                if existing_pythonpath:
                    # Prepend bundle_path to existing PYTHONPATH
                    env_vars["PYTHONPATH"] = f"{existing_pythonpath}{os.pathsep}{self._bundle_path}"
```

I aslo think we should add the bundle after any explicit user specified path

ashb commented on providers/standard/src/airflow/providers/standard/operators/python.py:
```diff
@@ -488,8 +488,30 @@ def execute(self, context: Context) -> Any:
+        if ti and hasattr(ti, "bundle_instance"):
+            bundle_instance = ti.bundle_instance
+            if bundle_instance and hasattr(bundle_instance, "path"):
+                return Path(bundle_instance.path)
```
Why bother makin it a path object when all we do is stringify it

ashb commented on providers/standard/src/airflow/providers/standard/operators/python.py:
```diff
@@ -565,6 +587,18 @@ def _execute_python_callable_in_subprocess(self, python_path: Path):
 
+            # Add bundle_path to PYTHONPATH for subprocess to import DAG bundle modules
+            if hasattr(self, "_bundle_path") and self._bundle_path:
+                bundle_path_str = str(self._bundle_path)
```
```suggestion
```

gfreitash committed:
Apply review suggestions

- Fix Dag nomenclature usage
- Simplify bundle path obtention logic
- Add bunde path after user-defined paths
- Use tmp_path pytest fixture instead of a manual one

jason810496 reviewed (commented):
> should I mark the reviews as resolved myself or the reviewer or other maintained will do that?

Thanks for the update! Please feel free to resolve the addressed comments.

Lee-W reviewed (approved):
Looks good. Just rebased. let's see whether there's any other failure

gfreitash committed:
Fix bundle path handling in subprocess operators

- Return `None` if `bundle_instance` is not present to avoid attribute errors.
- Ensure the bundle path is only added to `PYTHONPATH` when defined

gfreitash committed:
Refactor DAG bundle import test

- Mock subprocess execution to avoid environment-related issues.
- Test if the bundle path was passed to the subprocess in the env argument instead

gfreitash committed:
Remove unused pytest marks.

jscheffl reviewed (approved):
Looks good in my view.
